### PR TITLE
Refactor setup_training and remove test_mode

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Clean up and split `setup_training` into `setup_trainer` & `setup_training` to avoid call to `on_pretrain_routine_*` while testing. Removed redundant `test_mode`, use `trainer.testing` instead. Also a few minor updates, to make sure tests passes locally.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/cpu_accelerator.py [hunk 1]
```diff
@@ -53,10 +53,8 @@ def setup(self, model):
         self.trainer.model = model
 
     def train(self):
-        model = self.trainer.model
-
-        # set up training routine
-        self.trainer.train_loop.setup_training(model)
+        # set up trainer
```
Comment (carmocca) [inline pytorch_lightning/accelerators/cpu_accelerator.py:56]: this comment adds no value. same for the other accelerators

```suggestion
```

File: pytorch_lightning/accelerators/cpu_accelerator.py [hunk 2]
```diff
@@ -53,10 +53,8 @@ def setup(self, model):
         self.trainer.model = model
 
     def train(self):
```
Comment (carmocca) [inline pytorch_lightning/accelerators/cpu_accelerator.py:55]: why don't we define this function in `Accelerator`?

File: tests/models/test_hooks.py [hunk 3]
```diff
@@ -348,8 +348,6 @@ def on_test_model_train(self):
 
     expected = [
         'on_fit_start',
-        'on_pretrain_routine_start',
-        'on_pretrain_routine_end',
```
Comment (carmocca) [inline tests/models/test_hooks.py:352]: why are these removed?

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 4]
```diff
@@ -138,17 +134,23 @@ def setup(self, model, max_batches, dataloaders):
...
 
     def on_evaluation_epoch_start(self, *args, **kwargs):
-        if self.testing:
+        if self.trainer.testing:
             self.trainer.call_hook('on_test_epoch_start', *args, **kwargs)
         else:
             self.trainer.call_hook('on_validation_epoch_start', *args, **kwargs)
 
-    def build_args(self, test_mode, batch, batch_idx, dataloader_idx):
+    def build_args(self, batch, batch_idx, dataloader_idx):
```
Comment (Borda) [inline pytorch_lightning/trainer/evaluation_loop.py:142]: this will be a breaking change, pls keep it and add a warning for now+v0.2 about deprecation

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -64,7 +64,7 @@ def restore_weights(self, model: LightningModule) -> None:
             rank_zero_info(f'restored hpc model from: {checkpoint_path}')
 
         # 2. Attempt to restore states from `resume_from_checkpoint` file
-        elif self.trainer.resume_from_checkpoint is not None and not self.trainer.testing:
+        elif self.trainer.resume_from_checkpoint is not None:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:67]: why is this gone? thomas just added it in #5161

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -605,7 +643,7 @@ def run_evaluation(self, test_mode: bool = False, max_batches=None):
 
                 # lightning module methods
                 with self.profiler.profile("evaluation_step_and_end"):
-                    output = self.evaluation_loop.evaluation_step(test_mode, batch, batch_idx, dataloader_idx)
+                    output = self.evaluation_loop.evaluation_step(batch, batch_idx, dataloader_idx)
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:646]: also API chnage

File: tests/models/test_hooks.py (hunk 3 above)
Comment (rohitgr7) [reply tests/models/test_hooks.py:352]: I believe it's meant to be called during training only.
https://github.com/PyTorchLightning/pytorch-lightning/blob/4c6f36e6e14a5e3bace1fe32505ae0fe6f8bc682/pytorch_lightning/core/hooks.py#L87-L96

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/evaluation_loop.py:142]: isn't this an internal API change?

**Comment by Borda** [comment]
probably as a followup I would add two more refactoring:
1. rename `testing` to `is_testing` to be more intuaitive
2. add internal phase for the trainer as enum for Train/Valid/Test as we seem to have `LoggerStages`

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/trainer/evaluation_loop.py:142]: if so let's make it as protected `_ build_args` so anyone using it will crash rather on the name than bad arguments which are harder to spot...

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:67]: `self.trainer.checkpoint_connector.restore_weights(model)` will be called during training only.. so this won't be called while testing or evaluating. You can check the call to `setup_training` in the changes.

File: pytorch_lightning/accelerators/cpu_accelerator.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/cpu_accelerator.py:55]: good point :+1:

File: pytorch_lightning/accelerators/cpu_accelerator.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/cpu_accelerator.py:56]: yeah.. will update

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:646]: this one too? `_evaluation_step`?

**Comment by rohitgr7** [comment]
> probably as a followup I would add two more refactoring:
> 
>     1. rename `testing` to `is_testing` to be more intuaitive
> 
>     2. add internal phase for the trainer as enum for Train/Valid/Test as we seem to have `LoggerStages`


1. There will be an upcoming addition of `training.validate()` API, so this will be updated there.
2. Might need some help with this. Not familiar with Enum.

File: tests/models/test_hooks.py (hunk 3 above)
Comment (carmocca) [reply tests/models/test_hooks.py:352]: You are right, but was the test failing before? :thinking:

**Comment by Borda** [comment]
> 2\. Might need some help with this. Not familiar with Enum.

sure, just ping me :]

File: tests/models/test_hooks.py (hunk 3 above)
Comment (rohitgr7) [reply tests/models/test_hooks.py:352]: no, because these hooks were called during testing and the tests were expecting these hooks to be called.

Comment (rohitgr7) [reply tests/models/test_hooks.py:352]: but now they aren't

**Comment by carmocca** [comment]
>     2. add internal phase for the trainer as enum for Train/Valid/Test as we seem to have `LoggerStages`

This was also proposed in the `trainer.validate` PR (https://github.com/PyTorchLightning/pytorch-lightning/pull/4948#discussion_r534937487)

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
@rohitgr7 regarding accelerators, please have a look at #4510 #5385. we are working on some large scale refactoring.

**Comment by rohitgr7** [comment]
@awaelchli I see. Will remove other changes in the accelerators except `setup_training -> setup_trainer`. Will that be ok?

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -411,6 +411,48 @@ def __init__(
...
+
+        # set the ranks and devices
+        self.accelerator_backend.dist.rank = self.global_rank
+        self.accelerator_backend.dist.device = ref_model.device
+
+        # set local properties on the model
+        self.model_connector.copy_trainer_model_properties(ref_model)
+
+        # init amp. Must be done here instead of __init__ to allow ddp to work
+        if self.amp_backend == AMPType.NATIVE and self.precision == 16 and not self.use_tpu:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:436]: Not for this PR, but this should be moved into `AMPNativePlugin`.

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -125,46 +125,15 @@ def setup_fit(self, model, train_dataloader, val_dataloaders, datamodule):
...
-        Args:
-            model: The model to run sanity test on.
+    def setup_training(self):
         """
-        # --------------------------
-        # Setup??
-        # --------------------------
+        Sanity check a few things before starting actual training.
+        """
+        model = self.trainer.model
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:132]: Let's use model or `model_ref` directly. Not need to do `ref_model = model`

File: pytorch_lightning/accelerators/accelerator.py [hunk 9]
```diff
@@ -66,6 +70,7 @@ def train_or_test(self):
         if self.trainer.testing:
             results = self.trainer.run_test()
         else:
+            self.trainer.train_loop.setup_training()
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:73]: Much cleaner !

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -411,6 +411,48 @@ def __init__(
...
+            self.scaler = self.precision_connector.backend.scaler
+
+        # log hyper-parameters
+        if self.logger is not None:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:440]: Not for this PR, but this could be moved into its own function.

**Review by tchaton — COMMENTED** [review]
Great work ! Could you check `self.logger_connector.set_stage`. It is duplicating the information. 
Would be better to have a higher level State for this.
`self.logger_connector.set_stage` was introduced as validation can be performed during training and need to change store.

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:436]: I think Adrian is working on more refactors in a separate PR.

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:132]: I don't think that would be backward compatible, since model on cpu_accelerator has no ref_model, and we need ref_model for summarization and for calling hooks.

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:440]: yeah maybe to logger connector.

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:132]: done.

**Comment by rohitgr7** [comment]
> add internal phase for the trainer as enum for Train/Valid/Test as we seem to have LoggerStages

@Borda need help with this

**Comment by Borda** [comment]
> > add internal phase for the trainer as enum for Train/Valid/Test as we seem to have LoggerStages
> 
> @Borda need help with this

I made a draft for it here #5419 🐰

**Review by s-rog — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by rohitgr7** [comment]
@SeanNaren @tchaton mind review??

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -412,6 +412,46 @@ def __init__(
         # Callback system
         self.on_init_end()
 
+    def setup_trainer(self, model: LightningModule):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:415]: I think this refactor broke `trainer.test(resume_from_checkpoint="some_path", ...)`
I don't see where `self.trainer.checkpoint_connector.restore_weights()` is called in the evaluation loop or in the trainer. before this happened as part of the TrainLoop's `setup_training`

In `Trainer`'s `run_evaluation` I don't see any calls to the checkpoint connector either.

Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:415]: this is done in `setup_training` here: https://github.com/PyTorchLightning/pytorch-lightning/pull/5388/files#diff-6b21474ed45079f01dfa45ee3b9d40d23efe693c9c005ce897e25384ef425349

as of now restore is done while calling `.fit`. Previously it was happening during `.test()` too and was reloading the `resume_from_checkpoint` state and model weights that were wrong. But for your use-case, I opened an issue for reloading the callback states optionally.
